libcudacxx/include/cuda/__floating_point/fp.h

```diff
@@ -0,0 +1,530 @@
+}
+
+template <class _Config>
+class __fp
+{
+public:
+  template <class>
+  friend class __fp;
+  friend _Config;
+  friend struct __fp_cast;
+  friend struct __fp_ops;
+
+  using __config_type = _Config;
+
+  static constexpr size_t __exp_nbits  = __config_type::__exp_nbits;
+  static constexpr size_t __mant_nbits = __config_type::__mant_nbits;
+  static constexpr bool __is_signed    = __config_type::__is_signed;
+  static constexpr size_t __nbits      = __exp_nbits + __mant_nbits + __is_signed;
+  static constexpr bool __has_inf      = __config_type::__has_inf;
+  static constexpr bool __has_nan      = __config_type::__has_nan;
+  static constexpr bool __has_nans     = __config_type::__has_nans;
+  static constexpr bool __has_denorm   = __config_type::__has_denorm;
+
+  using __storage_type       = __fp_storage_t<__nbits>;
+  using __host_native_type   = typename __config_type::__host_native_type;
+  using __device_native_type = typename __config_type::__device_native_type;
+
+  static constexpr bool __has_host_native_type = !_CUDA_VSTD::is_same_v<__host_native_type, __fp_no_native_type_tag>;
+  static constexpr bool __has_device_native_type =
+    !_CUDA_VSTD::is_same_v<__device_native_type, __fp_no_native_type_tag>;
+
+  static_assert(!__has_host_native_type || sizeof(__storage_type) == sizeof(__host_native_type));
+  static_assert(!__has_device_native_type || sizeof(__storage_type) == sizeof(__device_native_type));
+
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr __fp __min()
+  {
+    return __config_type::template __min<__fp>();
+  }
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr __fp __max()
+  {
+    return __config_type::template __max<__fp>();
+  }
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr __fp __lowest()
+  {
+    __fp __ret{};
+    if constexpr (__is_signed)
+    {
+      __ret = __max();
+      __ret.__set_sign(true);
+    }
+    return __ret;
+  }
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr __fp __epsilon()
+  {
+    // TODO: implement epsilon
+    return __fp{};
+  }
+  _CCCL_TEMPLATE(bool _HasInf = __has_inf)
+  _CCCL_REQUIRES(_HasInf)
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr __fp __inf()
+  {
+    return __config_type::template __inf<__fp>();
+  }
+  _CCCL_TEMPLATE(bool _HasNan = __has_nan)
+  _CCCL_REQUIRES(_HasNan)
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr __fp __nan()
+  {
+    return __config_type::template __nan<__fp>();
+  }
+  _CCCL_TEMPLATE(bool _HasNans = __has_nans)
+  _CCCL_REQUIRES(_HasNans)
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr __fp __nans()
+  {
+    return __config_type::template __nans<__fp>();
+  }
+  _CCCL_TEMPLATE(bool _HasDenorm = __has_denorm)
+  _CCCL_REQUIRES(_HasDenorm)
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr __fp __denorm_min()
+  {
+    return __fp{__fp_from_storage, __storage_type{1}};
+  }
+
+  __fp() = default;
```

Comment:
We want to annotate defaulted SMF with `_CCCL_HIDE_FROM_ABI` also there were some issues with older toolchains, we should add `constexpr` to all of them because sometimes it is not properly deduced